audit/audit_upload_bucket.py

```diff
@@ -0,0 +1,274 @@
+    - If there are discrepencies, check if the file name and size is the same in the bucket as it is in metamist
+      - If the name and file size are the same, assume this file has just been moved around in the bucket
+4. Check if a sample has a completed cram - if so, its sequence data can be removed
+5. Any remaining sequence data in the bucket might require ingestion
+6. Create reports in the audit_results folder of the upload bucket, containing sequence files to delete,
+   ingest, and any samples without completed crams.
+"""
+
+
+from datetime import datetime
+import sys
+import logging
+import os
+import asyncio
+import click
+
+from cpg_utils.config import get_config
+
+from audit.generic_auditor import GenericAuditor
+
+FASTQ_EXTENSIONS = ('.fq.gz', '.fastq.gz', '.fq', '.fastq')
+BAM_EXTENSIONS = ('.bam',)
+CRAM_EXTENSIONS = ('.cram',)
+READ_EXTENSIONS = FASTQ_EXTENSIONS + BAM_EXTENSIONS + CRAM_EXTENSIONS
+GVCF_EXTENSIONS = ('.g.vcf.gz',)
+VCF_EXTENSIONS = ('.vcf', '.vcf.gz')
+ALL_EXTENSIONS = (
+    FASTQ_EXTENSIONS
+    + BAM_EXTENSIONS
+    + CRAM_EXTENSIONS
+    + GVCF_EXTENSIONS
+    + VCF_EXTENSIONS
+)
+
+FILE_TYPES_MAP = {
+    'fastq': FASTQ_EXTENSIONS,
+    'bam': BAM_EXTENSIONS,
+    'cram': CRAM_EXTENSIONS,
+    'all_reads': READ_EXTENSIONS,
+    'gvcf': GVCF_EXTENSIONS,
+    'vcf': VCF_EXTENSIONS,
+    'all': ALL_EXTENSIONS,
+}
+
+SEQUENCE_TYPES_MAP = {
+    'genome': [
+        'genome',
+    ],
+    'exome': [
+        'exome',
+    ],
+    'all': [
+        'genome',
+        'exome',
+    ],
+}
+
+
+class UploadBucketAuditor(GenericAuditor):
+    """Auditor specifically for upload cloud storage buckets"""
+
+    def __init__(
+        self,
+        dataset: str,
+        sequence_type: list[str],
+        file_types: tuple[str],
+        default_analysis_type='cram',
+        default_analysis_status='completed',
+    ):
+        super().__init__(
+            dataset=dataset,
+            sequence_type=sequence_type,
+            file_types=file_types,
+            default_analysis_type=default_analysis_type,
+            default_analysis_status=default_analysis_status,
+        )
+
+    def write_upload_bucket_audit_reports(
+        self,
+        access_level: str,
+        sequence_type_str: str,
+        file_types_str: str,
+        sequence_files_to_delete: list[tuple[str, int, str, list[int]]],
+        sequence_files_to_ingest: list[tuple[str, str, str, int, str]],
+        incomplete_samples: list[tuple[str, str]],
+    ):
+        """Write the sequences to delete and ingest to csv files and upload them to the bucket"""
+        # Writing the output to files with the file type, sequence type, and date specified
+        today = datetime.today().strftime('%Y-%m-%d')
+
+        bucket_name = f'cpg-{self.dataset}-main-upload'
+        if access_level == 'test':
+            bucket_name = f'cpg-{self.dataset}-upload'
```

Comment:
If this runs in Analysis-Runner you can probably swap this all out for 
`get_config()['storage']['default']['upload']`